lib/modules/datasource/generic.ts

```diff
@@ -0,0 +1,101 @@
+import is from '@sindresorhus/is';
+import jsonata from 'jsonata';
+import { z } from 'zod';
+import { logger } from '../../logger';
+import { readLocalFile } from '../../util/fs';
+import { HttpError } from '../../util/http';
+import { parseUrl } from '../../util/url';
+import { Datasource } from './datasource';
+import type { ReleaseResult } from './types';
+
+export const ReleaseResultZod = z
+  .object({
+    releases: z.array(
+      z
+        .object({
+          version: z.string(),
+          isDeprecated: z.boolean().optional(),
+          releaseTimestamp: z.string().optional(),
+          sourceUrl: z.string().optional(),
+          sourceDirectory: z.string().optional(),
+          changelogUrl: z.string().optional(),
+        })
+        .strict()
+    ),
+    sourceUrl: z.string().optional(),
+    sourceDirectory: z.string().optional(),
+    changelogUrl: z.string().optional(),
+    homepage: z.string().optional(),
+  })
+  .strict();
+
+export abstract class GenericDatasource extends Datasource {
+  override caching = true;
+  override customRegistrySupport = true;
+
+  protected async queryRegistry(
+    registryUrl: string | null | undefined
+  ): Promise<string | null> {
+    if (is.nullOrUndefined(registryUrl)) {
+      return null;
+    }
+
+    const url = parseUrl(registryUrl);
+    if (is.nullOrUndefined(url)) {
+      logger.debug(`Failed to parse url ${registryUrl}`);
+      return null;
+    }
+
+    let response: string | null = null;
+    try {
+      switch (url.protocol) {
+        case 'http:':
+        case 'https:':
+          response = (await this.http.get(registryUrl)).body;
+          break;
+        case 'file:':
+          response = await readLocalFile(
+            registryUrl.replace('file://', ''),
+            'utf8'
+          );
+          break;
```

Comment:
Does it make sense to use a repo local file as datasource for other files? 🤔  What are the use cases?

another issue:
If we use extract cache, then the repo isn't yet available, because the checkout is deferred until we know there is work todo.


So i would prefer to disable file based registry urls for now.